compiler/modules/nback/number.bal

```diff
@@ -156,41 +175,32 @@ function buildConvertToInt(llvm:Builder builder, Scaffold scaffold, bir:ConvertT
         buildConvertFloatToInt(builder, scaffold, val, insn);
         return;
     }
-    else if repr.base != BASE_REPR_TAGGED {
-        return scaffold.unimplementedErr("convert form decimal to int", insn.pos);
+    else if repr.base == BASE_REPR_TAGGED && repr.subtype == t:DECIMAL {
+        buildConvertDecimalToInt(builder, scaffold, val, insn);
+        return;
     }
     // convert to int form tagged pointer
+    buildStoreTagged(builder, scaffold, convertedInt(builder, scaffold, val, insn, convertToIntFunction), insn.result);
+}
 
-    llvm:PointerValue tagged = <llvm:PointerValue>val;
-    llvm:BasicBlock joinBlock = scaffold.addBasicBlock();
-
-    // semType must contain float or decimal. Since we don't have decimal yet in subset 6,
-    // it must contain float. In the future, below section is only needed conditionally.
-    llvm:Value hasType = buildHasTag(builder, tagged, TAG_FLOAT);
-    llvm:BasicBlock hasFloatBlock = scaffold.addBasicBlock();
-    llvm:BasicBlock noFloatBlock = scaffold.addBasicBlock();
-    builder.condBr(hasType, hasFloatBlock, noFloatBlock);
-    builder.positionAtEnd(hasFloatBlock);
-    buildConvertFloatToInt(builder, scaffold, buildUntagFloat(builder, scaffold, tagged), insn);
-    builder.br(joinBlock);
+function buildConvertFloatToInt(llvm:Builder builder, Scaffold scaffold, llvm:Value floatVal, bir:ConvertToIntInsn insn) {
+    buildStoreInt(builder, scaffold, convertedInt(builder, scaffold, floatVal, insn, floatToIntFunction), insn.result);
+}
 
-    builder.positionAtEnd(noFloatBlock);
-    buildStoreTagged(builder, scaffold, tagged, insn.result);
-    builder.br(joinBlock);
-    builder.positionAtEnd(joinBlock);
+function buildConvertDecimalToInt(llvm:Builder builder, Scaffold scaffold, llvm:Value decimalVal, bir:ConvertToIntInsn insn) {
+    buildStoreInt(builder, scaffold, convertedInt(builder, scaffold, decimalVal, insn, decimalToIntFunction), insn.result);
 }
 
-function buildConvertFloatToInt(llvm:Builder builder, Scaffold scaffold, llvm:Value floatVal, bir:ConvertToIntInsn insn) {
-    llvm:Value resultWithErr = <llvm:Value>builder.call(scaffold.getRuntimeFunctionDecl(floatToIntFunction), [floatVal]);
+function convertedInt(llvm:Builder builder, Scaffold scaffold, llvm:Value val, bir:ConvertToIntInsn insn, RuntimeFunction func) returns llvm:Value {
```

Comment:
Move the call to the RuntimeFunction out of this function. The job of this function then becomes to check the overflow, where the value can be either a tagged or int.  Call it something like buildCheckOverflow. It takes the resultWithErr as an argument, and returns the first member.

Reply:
Changed